src/runtime.js

```diff
@@ -509,6 +509,15 @@ export class MultidocManager {
       this.closeShadowRoot_(shadowRoot);
     };
 
+    /**
+     * Closes the document, resolving when visibility changes and services have
+     * been cleand up. The document can no longer be activated again.
+     * @return {Promise}
+     */
+    amp['closeAsync'] = () => {
+      return this.closeShadowRootAsync_(shadowRoot);
```

Comment:
I think this is a transparent enough change and we can just force `amp.close()` to become async, so there's no need for a new method.

Reply:
Just to be certain, do you envision fully async? Or run synchronous bits first, then return promise wrapping just the async bits? I mention this because the one use of `this.closeShadowRootAsync_` currently expects fully asynchronous behavior.

Reply:
Sorry, I keep noticing little extra bits to consider.
`attachShadowDoc_()` includes this near the start:
```
    if (shadowRoot.AMP) {
      user().warn(TAG, "Shadow doc wasn't previously closed");
      this.closeShadowRoot_(shadowRoot);
    }
```
It is not appropriate to return a promise from `attachShadowDoc_()` since this would be a breaking change for users expecting an object. We wouldn't be introducing a regression by making `.closeShadowRoot_()` return a promise, but we wouldn't be fixing all uses of it either.

Reply:
This is resolved as of e84ad45d321bc00ec7f9493ad4adb7d761f9e88a